In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
df=pd.read_csv(r"C:\Users\kashr\Documents\Own machine learning Project\employee_wellness_data.csv")

In [3]:
df.head()

Unnamed: 0  X1  age  gender  height  weight      steps  hear_rate  \
0           1   1   20       1   168.0    65.4  10.771429  78.531302   
1           2   2   20       1   168.0    65.4  11.475325  78.453390   
2           3   3   20       1   168.0    65.4  12.179221  78.540825   
3           4   4   20       1   168.0    65.4  12.883117  78.628260   
4           5   5   20       1   168.0    65.4  13.587013  78.715695   

    calories  distance  ...  entropy_setps  resting_heart  corr_heart_steps  \
0   0.344533  0.008327  ...       6.116349           59.0          1.000000   
1   3.287625  0.008896  ...       6.116349           59.0          1.000000   
2   9.484000  0.009466  ...       6.116349           59.0          1.000000   
3  10.154556  0.010035  ...       6.116349           59.0          1.000000   
4  10.825111  0.010605  ...       6.116349           59.0          0.982816   

   norm_heart  intensity_karvonen  sd_norm_heart  steps_times_distance  \
0   19.531302            0.138520       1.000000              0.089692   
1   19.453390            0.137967       1.000000              0.102088   
2   19.540825            0.138587       1.000000              0.115287   
3   19.628260            0.139208       1.000000              0.129286   
4   19.715695            0.139828       0.241567              0.144088   

        device activity mental_wellness  
0  apple watch    Lying               0  
1  apple watch    Lying               0  
2  apple watch    Lying               0  
3  apple watch    Lying               0  
4  apple watch    Lying               0  

[5 rows x 21 columns]

In [4]:
df.isnull().sum()


Unnamed: 0              0
X1                      0
age                     0
gender                  0
height                  0
weight                  0
steps                   0
hear_rate               0
calories                0
distance                0
entropy_heart           0
entropy_setps           0
resting_heart           0
corr_heart_steps        0
norm_heart              0
intensity_karvonen      0
sd_norm_heart           0
steps_times_distance    0
device                  0
activity                0
mental_wellness         0
dtype: int64

In [5]:
df.duplicated().sum()


0

In [6]:
df.shape

(6264, 21)

In [7]:
df=df.rename(columns={df.columns[7]:'heart_rate'})

In [8]:
df=df.drop(columns=['X1','Unnamed: 0'])

In [9]:
df.tail()

age  gender  height  weight  steps  heart_rate  calories  distance  \
6259   46       0   157.5    71.4    1.0        35.0      20.5       1.0   
6260   46       0   157.5    71.4    1.0        35.0      20.5       1.0   
6261   46       0   157.5    71.4    1.0        35.0      20.5       1.0   
6262   46       0   157.5    71.4    1.0        35.0      20.5       1.0   
6263   46       0   157.5    71.4    1.0        35.0      20.5       1.0   

      entropy_heart  entropy_setps  resting_heart  corr_heart_steps  \
6259            0.0            0.0           35.0               1.0   
6260            0.0            0.0           35.0               1.0   
6261            0.0            0.0           35.0               1.0   
6262            0.0            0.0           35.0               1.0   
6263            0.0            0.0           35.0               1.0   

      norm_heart  intensity_karvonen  sd_norm_heart  steps_times_distance  \
6259         0.0                 0.0            0.0                   1.0   
6260         0.0                 0.0            1.0                   1.0   
6261         0.0                 0.0            1.0                   1.0   
6262         0.0                 0.0            1.0                   1.0   
6263         0.0                 0.0            1.0                   1.0   

      device        activity  mental_wellness  
6259  fitbit  Running 7 METs                0  
6260  fitbit  Running 7 METs                0  
6261  fitbit  Running 7 METs                0  
6262  fitbit  Running 7 METs                0  
6263  fitbit  Running 7 METs                0

In [10]:
df['device'].value_counts()

device
apple watch    3656
fitbit         2608
Name: count, dtype: int64

In [11]:
df['device'].unique()

array(['apple watch', 'fitbit'], dtype=object)

In [12]:
df['activity'].unique()

array(['Lying', 'Sitting', 'Self Pace walk', 'Running 3 METs',
       'Running 5 METs', 'Running 7 METs'], dtype=object)

In [13]:
df.columns

Index(['age', 'gender', 'height', 'weight', 'steps', 'heart_rate', 'calories',
       'distance', 'entropy_heart', 'entropy_setps', 'resting_heart',
       'corr_heart_steps', 'norm_heart', 'intensity_karvonen', 'sd_norm_heart',
       'steps_times_distance', 'device', 'activity', 'mental_wellness'],
      dtype='object')

In [14]:
df.sample()


age  gender  height  weight     steps  heart_rate  calories  distance  \
5378   29       0   163.0    61.0  3.274123   36.091367       1.0  4.467105   

      entropy_heart  entropy_setps  resting_heart  corr_heart_steps  \
5378       6.189825       6.189825      26.335224               1.0   

      norm_heart  intensity_karvonen  sd_norm_heart  steps_times_distance  \
5378    9.756143            0.059249       2.272169             14.625851   

      device activity  mental_wellness  
5378  fitbit  Sitting                0

In [15]:
#  it is the library that I use For EDA purpose, because it gives the insights about the data in html file and it gives pretty good insights
from ydata_profiling import ProfileReport

In [16]:
prof=ProfileReport(df)
prof.to_file(output_file='output.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [17]:
df.head()

age  gender  height  weight      steps  heart_rate   calories  distance  \
0   20       1   168.0    65.4  10.771429   78.531302   0.344533  0.008327   
1   20       1   168.0    65.4  11.475325   78.453390   3.287625  0.008896   
2   20       1   168.0    65.4  12.179221   78.540825   9.484000  0.009466   
3   20       1   168.0    65.4  12.883117   78.628260  10.154556  0.010035   
4   20       1   168.0    65.4  13.587013   78.715695  10.825111  0.010605   

   entropy_heart  entropy_setps  resting_heart  corr_heart_steps  norm_heart  \
0       6.221612       6.116349           59.0          1.000000   19.531302   
1       6.221612       6.116349           59.0          1.000000   19.453390   
2       6.221612       6.116349           59.0          1.000000   19.540825   
3       6.221612       6.116349           59.0          1.000000   19.628260   
4       6.221612       6.116349           59.0          0.982816   19.715695   

   intensity_karvonen  sd_norm_heart  steps_times_distance       device  \
0            0.138520       1.000000              0.089692  apple watch   
1            0.137967       1.000000              0.102088  apple watch   
2            0.138587       1.000000              0.115287  apple watch   
3            0.139208       1.000000              0.129286  apple watch   
4            0.139828       0.241567              0.144088  apple watch   

  activity  mental_wellness  
0    Lying                0  
1    Lying                0  
2    Lying                0  
3    Lying                0  
4    Lying                0

In [18]:
df['mental_wellness'].value_counts()

mental_wellness
0    3252
1    3012
Name: count, dtype: int64

In [19]:
df.columns

Index(['age', 'gender', 'height', 'weight', 'steps', 'heart_rate', 'calories',
       'distance', 'entropy_heart', 'entropy_setps', 'resting_heart',
       'corr_heart_steps', 'norm_heart', 'intensity_karvonen', 'sd_norm_heart',
       'steps_times_distance', 'device', 'activity', 'mental_wellness'],
      dtype='object')

In [20]:
df.describe()

age       gender       height       weight        steps  \
count  6264.000000  6264.000000  6264.000000  6264.000000  6264.000000   
mean     29.158525     0.476533   169.709052    69.614464   109.562268   
std       8.908978     0.499489    10.324698    13.451878   222.797908   
min      18.000000     0.000000   143.000000    43.000000     1.000000   
25%      23.000000     0.000000   160.000000    60.000000     5.159534   
50%      28.000000     0.000000   168.000000    68.000000    10.092029   
75%      33.000000     1.000000   180.000000    77.300000   105.847222   
max      56.000000     1.000000   191.000000   115.000000  1714.000000   

        heart_rate     calories     distance  entropy_heart  entropy_setps  \
count  6264.000000  6264.000000  6264.000000    6264.000000    6264.000000   
mean     86.142331    19.471823    13.832555       6.030314       5.739984   
std      28.648385    27.309765    45.941437       0.765574       1.256348   
min       2.222222     0.056269     0.000440       0.000000       0.000000   
25%      75.598079     0.735875     0.019135       6.108524       5.909440   
50%      77.267680     4.000000     0.181719       6.189825       6.157197   
75%      95.669118    20.500000    15.697188       6.247928       6.247928   
max     194.333333    97.500000   335.000000       6.475733       6.475733   

       resting_heart  corr_heart_steps   norm_heart  intensity_karvonen  \
count    6264.000000       6264.000000  6264.000000         6264.000000   
mean       65.869938          0.306447    20.272393            0.155479   
std        21.203017          0.775418    28.388116            0.210927   
min         3.000000         -1.000000   -76.000000           -2.714286   
25%        58.134333         -0.467303     1.148883            0.009819   
50%        75.000000          0.665829     9.820254            0.079529   
75%        76.138701          1.000000    27.077336            0.211868   
max       155.000000          1.000000   156.319444            1.297980   

       sd_norm_heart  steps_times_distance  mental_wellness  
count    6264.000000           6264.000000      6264.000000  
mean        8.110854            590.035239         0.480843  
std        12.535080           4063.838530         0.499673  
min         0.000000              0.000690         0.000000  
25%         0.264722              0.659260         0.000000  
50%         2.893503             13.368619         0.000000  
75%         9.679672             93.728562         1.000000  
max        74.457929          51520.000000         1.000000

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6264 entries, 0 to 6263
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   age                   6264 non-null   int64  
 1   gender                6264 non-null   int64  
 2   height                6264 non-null   float64
 3   weight                6264 non-null   float64
 4   steps                 6264 non-null   float64
 5   heart_rate            6264 non-null   float64
 6   calories              6264 non-null   float64
 7   distance              6264 non-null   float64
 8   entropy_heart         6264 non-null   float64
 9   entropy_setps         6264 non-null   float64
 10  resting_heart         6264 non-null   float64
 11  corr_heart_steps      6264 non-null   float64
 12  norm_heart            6264 non-null   float64
 13  intensity_karvonen    6264 non-null   float64
 14  sd_norm_heart         6264 non-null   float64
 15  steps_times_distance 

# Calculate BMI 

In [22]:
df['BMI'] = df['weight'] / ((df['height'] / 100) ** 2)

In [24]:
# now dropping height and weight column as it will be included in BMI itself (feature extraction and selection)

In [23]:
df.drop(columns=['weight','height'],inplace=True)

In [27]:
# Encoding categorical values
df=pd.get_dummies(df,columns=['device','activity'],drop_first=True).astype(int)

In [28]:
df.sample()

age  gender  steps  heart_rate  calories  distance  entropy_heart  \
1623   31       1    130         102         0         0              5   

      entropy_setps  resting_heart  corr_heart_steps  ...  sd_norm_heart  \
1623              5             66                 0  ...              9   

      steps_times_distance  mental_wellness  BMI  device_fitbit  \
1623                    13                0   28              0   

      activity_Running 3 METs  activity_Running 5 METs  \
1623                        0                        1   

      activity_Running 7 METs  activity_Self Pace walk  activity_Sitting  
1623                        0                        0                 0  

[1 rows x 22 columns]

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6264 entries, 0 to 6263
Data columns (total 22 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   age                      6264 non-null   int32
 1   gender                   6264 non-null   int32
 2   steps                    6264 non-null   int32
 3   heart_rate               6264 non-null   int32
 4   calories                 6264 non-null   int32
 5   distance                 6264 non-null   int32
 6   entropy_heart            6264 non-null   int32
 7   entropy_setps            6264 non-null   int32
 8   resting_heart            6264 non-null   int32
 9   corr_heart_steps         6264 non-null   int32
 10  norm_heart               6264 non-null   int32
 11  intensity_karvonen       6264 non-null   int32
 12  sd_norm_heart            6264 non-null   int32
 13  steps_times_distance     6264 non-null   int32
 14  mental_wellness          6264 non-null   int32
 15  BMI 

In [30]:
X=df.drop(columns='mental_wellness')
y=df['mental_wellness']

# Applying Pca

In [35]:
from sklearn.decomposition import PCA
# Step 1: Standardize the data
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Step 2: Apply PCA
pca = PCA(n_components=5)  # Reducing to 5 principal components
X_pca = pca.fit_transform(X_scaled)

In [36]:
from sklearn.model_selection import train_test_split

In [44]:
X_train_scaled,X_test_scaled,y_train,y_test=train_test_split(X_pca,y,random_state=42,test_size=0.2)

In [45]:
X_train_scaled

array([[-2.50879555, -3.0638705 ,  1.38718013,  2.32724875,  0.12402319],
       [-0.52047019, -0.51068142, -0.60962528,  0.63071979, -1.44005982],
       [-4.73118559, -0.93889992,  1.5407562 ,  2.98773831,  1.21862953],
       ...,
       [ 0.39076261, -0.79927655,  1.41480499, -0.408976  ,  0.79767619],
       [-0.25791174,  1.67889235, -0.80895295,  1.28412407, -1.8334344 ],
       [ 1.05205082, -0.13371555, -0.25050378,  0.66924612, -0.68486902]])

In [46]:
y_train

2024    1
3277    0
5015    1
2572    1
1024    1
       ..
3772    1
5191    0
5226    0
5390    1
860     1
Name: mental_wellness, Length: 5011, dtype: int32

In [47]:
X_test_scaled

array([[-0.02269958, -1.94553644, -0.14448283, -0.85084624, -0.67785425],
       [ 0.62775761, -1.4724159 , -0.18985395,  0.06536964,  0.41397723],
       [ 0.60163105,  2.52991417,  1.76823878,  0.47441924,  1.57060618],
       ...,
       [ 1.22887472, -0.83625186, -0.26525245, -1.39033298,  0.96093759],
       [-0.29999   , -2.68081761,  1.01615242,  1.94206864, -0.18123884],
       [ 0.76432651,  1.13564542, -1.2138987 , -0.40480097, -0.53959734]])

In [48]:
from sklearn.linear_model import LogisticRegression

In [49]:
lr_model=LogisticRegression(penalty='l2')

In [50]:
lr_model.fit(X_train_scaled,y_train)

LogisticRegression()

In [51]:
y_pred=lr_model.predict(X_test_scaled)

In [52]:
from sklearn.metrics import accuracy_score

In [53]:
accuracy_score(y_test,y_pred)

0.690343176376696

In [54]:
# Lets check the accuracy score on training data 

In [55]:
X_train_prediction=lr_model.predict(X_train_scaled)
training_data_accuracy=accuracy_score(X_train_prediction,y_train)
training_data_accuracy

0.6958690880063859

In [ ]:
#  Lets check using random forest 

In [92]:
from sklearn.ensemble import RandomForestClassifier
rf_model=RandomForestClassifier(n_estimators=200,max_samples=0.8,min_samples_leaf=1,min_samples_split=2,random_state=42)

In [104]:
rf_model.fit(X_train_scaled,y_train)

RandomForestClassifier(max_samples=0.8, n_estimators=200, random_state=42)

In [107]:
y_pred_rf=rf_model.predict(X_test_scaled)

In [108]:
accuracy_score(y_test,y_pred_rf)

0.951316839584996

In [109]:
from sklearn.neighbors import KNeighborsClassifier

In [110]:
knn_model=KNeighborsClassifier()

In [111]:
knn_model.fit(X_train_scaled,y_train)

KNeighborsClassifier()

In [112]:
knn_pred=knn_model.predict(X_test_scaled)

In [113]:
accuracy_score(y_test,knn_pred)

0.9441340782122905

In [114]:
from xgboost import XGBClassifier

In [115]:
xgb_model= XGBClassifier(booster="gbtree",learning_rate=0.1,max_depth=6,min_child_weight=1,subsample=0.8,colsample_bytree=0.8,gamma=0,reg_lambda=1,reg_alpha=0,objective="binary:logistic",eval_metric="logloss")

In [116]:
xgb_model.fit(X_train_scaled,y_train)

XGBClassifier(base_score=None, booster='gbtree', callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=0.8, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='logloss',
              feature_types=None, gamma=0, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=0.1, max_bin=None, max_cat_threshold=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=6,
              max_leaves=None, min_child_weight=1, missing=nan,
              monotone_constraints=None, multi_strategy=None, n_estimators=None,
              n_jobs=None, num_parallel_tree=None, random_state=None, ...)

In [117]:
xgb_pred=xgb_model.predict(X_test_scaled)

In [118]:
accuracy_score(y_test,xgb_pred)

0.9241819632881085

# So After evaluating and applying cross validation(randomly splitting the dataset 10 times), Logistic gives accuracy of 69.5% and random forest gives 95% accuracy and knn gives 94% accuracy and XGboost is giving 92.4% accuracy

In [119]:
from sklearn.ensemble import VotingClassifier

# So we will select our best three model and apply voting ensemble to further improve model's overall performance and generalise our model

In [125]:
vc=VotingClassifier(estimators=[('rf_model',rf_model),('xgb_model',xgb_model),('knn_model',knn_model)])

In [126]:
vc.fit(X_train_scaled,y_train)

VotingClassifier(estimators=[('rf_model',
                              RandomForestClassifier(max_samples=0.8,
                                                     n_estimators=200,
                                                     random_state=42)),
                             ('xgb_model',
                              XGBClassifier(base_score=None, booster='gbtree',
                                            callbacks=None,
                                            colsample_bylevel=None,
                                            colsample_bynode=None,
                                            colsample_bytree=0.8, device=None,
                                            early_stopping_rounds=None,
                                            enable_categorical=False,
                                            eval_metric='logloss',
                                            feature_...
                                            importance_type=None,
                                            interaction_constraints=None,
                                            learning_rate=0.1, max_bin=None,
                                            max_cat_threshold=None,
                                            max_cat_to_onehot=None,
                                            max_delta_step=None, max_depth=6,
                                            max_leaves=None, min_child_weight=1,
                                            missing=nan,
                                            monotone_constraints=None,
                                            multi_strategy=None,
                                            n_estimators=None, n_jobs=None,
                                            num_parallel_tree=None,
                                            random_state=None, ...)),
                             ('knn_model', KNeighborsClassifier())])

In [127]:
vc_pred=vc.predict(X_test_scaled)

In [128]:
accuracy_score(y_test,vc_pred)

0.9489225857940942

# So at last after combing these model we can confidently say our model is giving 94.89% accuracy

In [130]:
import pickle

with open("Employee_wellness_prediction.pkl", "wb") as file:
    pickle.dump(vc, file)
